https://www.kaggle.com/datasets/nikhil1e9/loan-default 

# Data Collection

In this section, we will load and join the datasets. 

In [1]:
import numpy as np
import pandas as pd

In [2]:
loans = pd.read_csv('loan_default.csv')

In [3]:
loans.head(3)

LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0  I38PQUQS96   56   85994       50587          520              80   
1  HPSK72WA7R   69   50432      124440          458              15   
2  C1OZ6DPJ8Y   46   84208      129188          451              26   

   NumCreditLines  InterestRate  LoanTerm  DTIRatio   Education  \
0               4         15.23        36      0.44  Bachelor's   
1               1          4.81        60      0.68    Master's   
2               3         21.17        24      0.31    Master's   

  EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
0      Full-time      Divorced         Yes           Yes       Other   
1      Full-time       Married          No            No       Other   
2     Unemployed      Divorced         Yes           Yes        Auto   

  HasCoSigner  Default  
0         Yes        0  
1         Yes        0  
2          No        1

Since we are only interested in the 'loan_defaults.csv' dataset, we do not need to use pandas merge, join, concatenate or compare methods. 

# Data Organization

In this section, we will organize the file structure for this project.

# Data Definition

Questions
    
    What are the column names?
    
    What are the column datatypes?
    
    What information does each column column contain?
    
    Do your column names correspond to what those columns store?

The following dataframe contains answers all of these questions.

In [4]:
columns = loans.columns.tolist()
column_dtypes = loans.dtypes
column_descriptions = [
    'A unique identifier for each loan', 'The age of the borrower', 'The annual income of the borrower', 
    'The amount of money being borrowed', 'The credit score of the borrower', 'The number of months the borrower has been employed', 
    'The number of credit lines the borrower has open', 'The interest rate for the loan', 'The term length of the loan in months', 
    'The Debt-to-Income ratio', 'The highest level of education attained by the borrower', 'The type of employment status of the borrower', 
    'The marital status of the borrower', 'Whether the borrower has a mortgage', 'Whether the borrower has dependents', 
    'The purpose of the loan', 'Whether the loan has a co-signer', 'Indicate whether the loan defaulted or not'
]
pd.DataFrame(data=list(zip(column_descriptions, column_dtypes)), 
             columns=['ColumnDescription', 'ColumnDatatype'], 
             index=columns)

ColumnDescription  \
LoanID                          A unique identifier for each loan   
Age                                       The age of the borrower   
Income                          The annual income of the borrower   
LoanAmount                     The amount of money being borrowed   
CreditScore                      The credit score of the borrower   
MonthsEmployed  The number of months the borrower has been emp...   
NumCreditLines   The number of credit lines the borrower has open   
InterestRate                       The interest rate for the loan   
LoanTerm                    The term length of the loan in months   
DTIRatio                                 The Debt-to-Income ratio   
Education       The highest level of education attained by the...   
EmploymentType      The type of employment status of the borrower   
MaritalStatus                  The marital status of the borrower   
HasMortgage                   Whether the borrower has a mortgage   
HasDependents                 Whether the borrower has dependents   
LoanPurpose                               The purpose of the loan   
HasCoSigner                      Whether the loan has a co-signer   
Default                Indicate whether the loan defaulted or not   

               ColumnDatatype  
LoanID                 object  
Age                     int64  
Income                  int64  
LoanAmount              int64  
CreditScore             int64  
MonthsEmployed          int64  
NumCreditLines          int64  
InterestRate          float64  
LoanTerm                int64  
DTIRatio              float64  
Education              object  
EmploymentType         object  
MaritalStatus          object  
HasMortgage            object  
HasDependents          object  
LoanPurpose            object  
HasCoSigner            object  
Default                 int64

Questions

    What are the counts and percents for each unique value?

In [5]:
categorical_columns = [
    'Education', 'EmploymentType', 'MaritalStatus', 'HasMortgage', 'HasDependents', 'LoanPurpose', 'HasCoSigner'
]

def categorical_count_and_percent_dataframe(column):
    unique_values = loans[column].unique().tolist()
    counts_and_proportions = list(
        zip(
            loans[column].value_counts().tolist(), 
            loans[column].value_counts(normalize=True).tolist()
        )
    )
    return pd.DataFrame(
        data = counts_and_proportions, 
        index = unique_values, 
        columns = ['Count', 'Percent']
    )

categorical_descriptions = [categorical_count_and_percent_dataframe(categorical_columns[i]) for i in range(len(categorical_columns))]

for i in range(len(categorical_descriptions)):
    print(categorical_descriptions[i])

             Count   Percent
Bachelor's   64366  0.252073
Master's     63903  0.250259
High School  63541  0.248842
PhD          63537  0.248826
               Count   Percent
Full-time      64161  0.251270
Unemployed     63824  0.249950
Self-employed  63706  0.249488
Part-time      63656  0.249292
          Count   Percent
Divorced  85302  0.334063
Married   85033  0.333010
Single    85012  0.332927
      Count   Percent
Yes  127677  0.500014
No   127670  0.499986
      Count   Percent
Yes  127742  0.500268
No   127605  0.499732
           Count   Percent
Other      51298  0.200895
Auto       51286  0.200848
Business   51005  0.199748
Home       50914  0.199391
Education  50844  0.199117
      Count   Percent
Yes  127701  0.500108
No   127646  0.499892


Unique values are relatively evenly proportioned.

In [6]:
loans_description = loans.describe().T
loans_description['range'] = loans_description['max'] - loans_description['min']
loans_description

count           mean           std      min       25%  \
Age             255347.0      43.498306     14.990258     18.0     31.00   
Income          255347.0   82499.304597  38963.013729  15000.0  48825.50   
LoanAmount      255347.0  127578.865512  70840.706142   5000.0  66156.00   
CreditScore     255347.0     574.264346    158.903867    300.0    437.00   
MonthsEmployed  255347.0      59.541976     34.643376      0.0     30.00   
NumCreditLines  255347.0       2.501036      1.117018      1.0      2.00   
InterestRate    255347.0      13.492773      6.636443      2.0      7.77   
LoanTerm        255347.0      36.025894     16.969330     12.0     24.00   
DTIRatio        255347.0       0.500212      0.230917      0.1      0.30   
Default         255347.0       0.116128      0.320379      0.0      0.00   

                      50%        75%       max     range  
Age                 43.00      56.00      69.0      51.0  
Income           82466.00  116219.00  149999.0  134999.0  
LoanAmount      127556.00  188985.00  249999.0  244999.0  
CreditScore        574.00     712.00     849.0     549.0  
MonthsEmployed      60.00      90.00     119.0     119.0  
NumCreditLines       2.00       3.00       4.0       3.0  
InterestRate        13.46      19.25      25.0      23.0  
LoanTerm            36.00      48.00      60.0      48.0  
DTIRatio             0.50       0.70       0.9       0.8  
Default              0.00       0.00       1.0       1.0

All columns are large enough to run simulations on the dataframe. 

The youngest borrower is 18; the oldest is 69. The mean borrower age is 43 with a standard deviation of 15 years. 

The minimum annual income for a borrower is \\$15,000. The maximum annual income for a borrower is been \\$150,000. The mean borrower incomes is \\$82,000 with a standard deviation of \\$39,000. 

The minimum loan is \\$5,000. The maximum loan is been \\$250,000. The mean loan amount is $128,000 with a standard deviation of \\$71,000. Therefore, the mean loan is almost twice the mean annual income. 

The minimum credit score for a borrower is 300; the highest is 850. The mean is about 575 with a standard deviation of 159. 

Some borrowers are not employed, while the longest a borrower has been employed is 10 years. The mean borrowers has been employed 5 years with a standard deviation of almost 3 years. 

"A line of credit (LOC) is a preset borrowing limit offered by banks and financial institutions to their personal and business customers. Lines of credit can be used at any time until the limit is reached. ([Investopedia](https://www.investopedia.com/terms/l/lineofcredit.asp))" Some borrowers have no credit lines. The most any borrower has is 4. The mean credit lines per borrower is 2.5 with a standard deviation of one LOC. 

The mean interest rate for a loan is 13% with a standard deviation of 7%. The losest interest rate for any loan is 2% while the maximum is 25%. 

The mean loan term is 3 years with a standard deviation of 17 months. The shortest loan term is a year. The maximum is 5 years. 

The debt to income ratio (DTI) measure the percent of a borrower's monthy income that goes towards paying off the loan. The mean DTI ratio is about 50% with a standard deviation of 0.23%. The lowest DTI ratio is 0.1%. The highest is 90%. 

`Income`, `LoanAmount`, `InterestRate`, and `LoanTerm` should determine `DTIRatio`. 

On average, 12% of loans end in a default. 

# Data Cleaning

The `DTIRatio` is equal to `MonthlyDue` divided by `MonthlyIncome`. 

On the one hand, `MonthlyDue` is equal to `LoanAmount` plus `Interest` divided by `LoanTerm`, 

where `Interest` is equal to `LoanAmount` times `InterestRate` times `LoanTerm`. 

On the other hand, `MonthlyIncome` is equal to `Income` divided by 12.

In [7]:
ratios = pd.DataFrame()

In [8]:
ratios[[
    'DTIRatio', 'Income', 'LoanAmount', 'LoanTerm'
]] = loans[[
    'DTIRatio', 'Income', 'LoanAmount', 'LoanTerm']]

ratios['DTIEstimatedDue'] = ratios['DTIRatio'] * ratios['Income'] * (ratios['LoanTerm'] / 12)

ratios.head()

DTIRatio  Income  LoanAmount  LoanTerm  DTIEstimatedDue
0      0.44   85994       50587        36        113512.08
1      0.68   50432      124440        60        171468.80
2      0.31   84208      129188        24         52208.96
3      0.23   31713       44799        24         14587.98
4      0.73   20437        9139        48         59676.04

In [13]:
ratios['AnnualInterestRate'] = loans['InterestRate'] / 100

ratios['Interest'] = ratios['LoanAmount'] * ratios['AnnualInterestRate'] * (ratios['LoanTerm'] / 12)

ratios['EstimatedDue'] = ratios['LoanAmount'] + ratios['Interest']

ratios.head()

DTIRatio  Income  LoanAmount  LoanTerm  DTIEstimatedDue  \
0      0.44   85994       50587        36        113512.08   
1      0.68   50432      124440        60        171468.80   
2      0.31   84208      129188        24         52208.96   
3      0.23   31713       44799        24         14587.98   
4      0.73   20437        9139        48         59676.04   

   AnnualInterestRate    Interest  EstimatedDue  
0              0.1523  23113.2003    73700.2003  
1              0.0481  29927.8200   154367.8200  
2              0.2117  54698.1992   183886.1992  
3              0.0707   6334.5786    51133.5786  
4              0.0651   2379.7956    11518.7956

Based on the `DTIRatio` and `LoanTerm` of row 2, the borrower will pay \\$52K on a \\$129K loan.

In [27]:
loans.isna().sum()

LoanID            0
Age               0
Income            0
LoanAmount        0
CreditScore       0
MonthsEmployed    0
NumCreditLines    0
InterestRate      0
LoanTerm          0
DTIRatio          0
Education         0
EmploymentType    0
MaritalStatus     0
HasMortgage       0
HasDependents     0
LoanPurpose       0
HasCoSigner       0
Default           0
dtype: int64

In [32]:
loans = loans.dropna()
loans.head()

LoanID  Age  Income  LoanAmount  CreditScore  MonthsEmployed  \
0  I38PQUQS96   56   85994       50587          520              80   
1  HPSK72WA7R   69   50432      124440          458              15   
2  C1OZ6DPJ8Y   46   84208      129188          451              26   
3  V2KKSFM3UN   32   31713       44799          743               0   
4  EY08JDHTZP   60   20437        9139          633               8   

   NumCreditLines  InterestRate  LoanTerm  DTIRatio    Education  \
0               4         15.23        36      0.44   Bachelor's   
1               1          4.81        60      0.68     Master's   
2               3         21.17        24      0.31     Master's   
3               3          7.07        24      0.23  High School   
4               4          6.51        48      0.73   Bachelor's   

  EmploymentType MaritalStatus HasMortgage HasDependents LoanPurpose  \
0      Full-time      Divorced         Yes           Yes       Other   
1      Full-time       Married          No            No       Other   
2     Unemployed      Divorced         Yes           Yes        Auto   
3      Full-time       Married          No            No    Business   
4     Unemployed      Divorced          No           Yes        Auto   

  HasCoSigner  Default  
0         Yes        0  
1         Yes        0  
2          No        1  
3          No        0  
4          No        0